File: c_src/dlcbf.c
/*
** -------------------------------------------------------------------
**
** dlcbf.c: d-left counting bloom filter implementation
**
** Copyright (c) 2012 Basho Technologies, Inc. All Rights Reserved.
**
** This file is provided to you under the Apache License,
** Version 2.0 (the "License"); you may not use this file
** except in compliance with the License.  You may obtain
** a copy of the License at
**
**   http:**www.apache.org/licenses/LICENSE-2.0
**
** Unless required by applicable law or agreed to in writing,
** software distributed under the License is distributed on an
** "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
** KIND, either express or implied.  See the License for the
** specific language governing permissions and limitations
** under the License.
**
** -------------------------------------------------------------------
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "dlcbf.h"

#if defined(__APPLE__)
#  define COMMON_DIGEST_FOR_OPENSSL
#  include <CommonCrypto/CommonDigest.h>
#  define SHA1 CC_SHA1
#else
#  include <openssl/sha.h>
#endif

#define HASH_BYTES 20


Dlcbf*
dlcbf_init(unsigned d, unsigned b)
{
    Dlcbf* dlcbf = malloc(sizeof(Dlcbf));
    *((unsigned*)&dlcbf->d) = d;
    *((unsigned*)&dlcbf->b) = b;
    *((unsigned*)&dlcbf->bits) = rint(log(b)/log(2));

    DlcbfTable* table = dlcbf->tables = malloc(sizeof(DlcbfTable)*d);
    DlcbfTable* tables_end = table + d;
    const size_t numbuckets = sizeof(DlcbfBucket)*b;
    while (table != tables_end) {
        table->buckets = malloc(numbuckets);
        memset(table++->buckets, 0, numbuckets);
    }
    return dlcbf;
}

void
dlcbf_destroy(Dlcbf* dlcbf)
{
    DlcbfTable* table = dlcbf->tables;
    DlcbfTable* tables_end = table + dlcbf->d;
    while (table != tables_end) {
        free(table++->buckets);
    }
    free(dlcbf->tables);
    free(dlcbf);
}

static unsigned
get_bits(const unsigned char* input, unsigned numbits, unsigned pos)
{
    unsigned value = 0;
    unsigned posbits = pos % 8;
    unsigned i = pos/8, start = posbits > numbits ? numbits : posbits;

    if (start != 0) {
        value = input[i++] & (0xFF >> (8 - start));
        numbits -= start;
    }
    while (numbits >= 8) {
        value = (value << 8) | input[i++];
        numbits -= 8;
    }
    if (numbits != 0) {
        unsigned char last = (input[i] >> (8 - numbits));
        value = (value << numbits) | last;
    }
    return value;
}

static DlcbfBucketFingerprint*
get_targets(unsigned d, unsigned bits, const unsigned char* hash, DlcbfBucketFingerprint* targets)
{
    unsigned pos;
    DlcbfBucketFingerprint* targets_end = targets + d;
    for (pos = 0; targets != targets_end; ++targets, pos += FINGERPRINT_BITSIZE) {
        targets->bucket_i = get_bits(hash, bits, pos);
        targets->fingerprint = get_bits(hash, FINGERPRINT_BITSIZE, pos += bits);
    }
    return targets;
}

static DlcbfField*
item_location(const Fingerprint fingerprint, DlcbfBucket* bucket)
{
    DlcbfField f;
    DlcbfField* field = bucket->fields;
    DlcbfField* fields_end = field + bucket->count;
    for (f.f.fingerprint = fingerprint; field != fields_end; ++field) {
        if (f.f.fingerprint == field->f.fingerprint) {
            return field;
        }
    }
    return NULL;
}

void
dlcbf_add(const unsigned char* data, unsigned length, Dlcbf* dlcbf)
{
    unsigned char hash[HASH_BYTES];
    SHA1(data, length, hash);

    DlcbfBucketFingerprint targets[dlcbf->d];
    get_targets(dlcbf->d, dlcbf->bits, hash, targets);

    DlcbfTable* table = dlcbf->tables;
    DlcbfTable* tables_end = table + dlcbf->d;

    DlcbfBucketFingerprint* target = targets;
    unsigned lowest_count = table->buckets[target->bucket_i].count;
    DlcbfBucket* bucket = &table->buckets[target->bucket_i];
    Fingerprint fingerprint = target->fingerprint;

    for (++target, ++table; table != tables_end; ++target, ++table) {
        unsigned count = table->buckets[target->bucket_i].count;
        if (count < lowest_count) {
            lowest_count = count;
            bucket = &table->buckets[target->bucket_i];
            fingerprint = target->fingerprint;
        }
    }

    bucket->fields[lowest_count].f.fingerprint = fingerprint;
    ++bucket->fields[lowest_count].f.count;
    ++bucket->count;
}

static DlcbfLoc
location_of(const unsigned char* data, unsigned length, Dlcbf* dlcbf, DlcbfBucket** bucket)
{
    unsigned char hash[HASH_BYTES];
    SHA1(data, length, hash);

    DlcbfBucketFingerprint targets[dlcbf->d];
    get_targets(dlcbf->d, dlcbf->bits, hash, targets);

    DlcbfBucketFingerprint* target = targets;
    DlcbfBucketFingerprint* targets_end = target + dlcbf->d;
    DlcbfTable* table = dlcbf->tables;
    DlcbfLoc loc = {NULL};
    while (target != targets_end) {
        *bucket = &table++->buckets[target->bucket_i];
        if ((loc.field = item_location(target++->fingerprint, *bucket)) != NULL) {
            break;
        }
    }
    if (loc.field == NULL) {
        *bucket = NULL;
    }
    return loc;
}

int
dlcbf_member(const unsigned char* data, unsigned length, Dlcbf* dlcbf)
{
    DlcbfBucket* bucket;
    const DlcbfLoc loc = location_of(data, length, dlcbf, &bucket);
    return loc.field != NULL;
}

void
dlcbf_delete(const unsigned char* data, unsigned length, Dlcbf* dlcbf)
{
    unsigned char hash[HASH_BYTES];
    SHA1(data, length, hash);
    DlcbfBucket* bucket;
    DlcbfLoc loc = location_of(data, length, dlcbf, &bucket);
    if (loc.field != NULL) {
        --bucket->count;
        --(loc.field->f.count);
    }
}

File: src/dlcbf.erl
%% -------------------------------------------------------------------
%%
%% dlcbf: Erlang NIF wrapper for d-left counting bloom filter
%%
%% Copyright (c) 2012 Basho Technologies, Inc. All Rights Reserved.
%%
%% This file is provided to you under the Apache License,
%% Version 2.0 (the "License"); you may not use this file
%% except in compliance with the License.  You may obtain
%% a copy of the License at
%%
%%   http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing,
%% software distributed under the License is distributed on an
%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
%% KIND, either express or implied.  See the License for the
%% specific language governing permissions and limitations
%% under the License.
%%
%% -------------------------------------------------------------------
-module(dlcbf).
-author("Ian Plosker").
-author("Steve Vinoski").

-export([new/2,
         add/2,
         delete/2,
         in/2]).

-ifdef(TEST).
-ifdef(EQC).
-include_lib("eqc/include/eqc.hrl").
-export([pos_int/0,
         power_of_two/0,
         prop_add_are_members/0,
         check_membership/4,
         keys/0,
         ops/1,
         apply_ops/3,
         prop_add_delete/0]).
-endif.
-include_lib("eunit/include/eunit.hrl").
-endif.

-on_load(init/0).

-define(nif_stub, nif_stub_error(?LINE)).
nif_stub_error(Line) ->
    erlang:nif_error({nif_not_loaded,module,?MODULE,line,Line}).

-spec init() -> ok | {error, any()}.
init() ->
    case code:priv_dir(dlht) of
        {error, bad_name} ->
            case code:which(?MODULE) of
                Filename when is_list(Filename) ->
                    SoName = filename:join([filename:dirname(Filename),"../priv", "dlcbf"]);
                _ ->
                    SoName = filename:join("../priv", "dlcbf")
            end;
         Dir ->
            SoName = filename:join(Dir, "dlcbf")
    end,
    erlang:load_nif(SoName, 0).

-spec new(pos_integer(), pos_integer()) -> {ok, reference()}.
new(_D, _B) ->
    ?nif_stub.

-spec add(binary(), reference()) -> ok.
add(_Bin, _Dlht) ->
    ?nif_stub.

-spec delete(binary(), reference()) -> ok.
delete(_Bin, _Dlht) ->
    ?nif_stub.

-spec in(binary(), reference()) -> boolean().
in(_Bin, _Dlht) ->
    ?nif_stub.


-ifdef(TEST).

basic_test() ->
    {ok, D} = new(2,16),
    ok = add(<<"a">>, D),
    ok = add(<<"b">>, D),
    ok = add(<<"c">>, D),
    ?assert(in(<<"a">>, D)),
    ?assert(in(<<"b">>, D)),
    ?assert(in(<<"c">>, D)),
    ?assertNot(in(<<"d">>, D)),
    ?assertNot(in(<<"e">>, D)),
    ?assertNot(in(<<"f">>, D)),
    ok = delete(<<"c">>, D),
    ?assertNot(in(<<"c">>, D)),
    ?assert(in(<<"a">>, D)),
    ?assert(in(<<"b">>, D)),
    ok = delete(<<"a">>, D),
    ?assertNot(in(<<"a">>, D)),
    ?assertNot(in(<<"c">>, D)),
    ?assert(in(<<"b">>, D)).

-ifdef(EQC).

-define(QC_OUT(P),
        eqc:on_output(fun(Str, Args) -> io:format(user, Str, Args) end, P)).

qc(P) ->
    ?assert(eqc:quickcheck(?QC_OUT(P))).

basic_quickcheck_test_() ->
    Prop = eqc:numtests(5000, dlcbf:prop_add_are_members()),
    {timeout, 2*60, fun() -> qc(Prop) end}.

pos_int() ->
    ?LET(N, int(), abs(N) + 1).

power_of_two() ->
    ?LET(N, pos_int(), begin trunc(math:pow(2, N)) end).

prop_add_are_members() ->
    ?FORALL(M, non_empty(list(binary())),
            ?FORALL(N, non_empty(list(binary())),
                    ?LET({D, B}, {4, 2048},
                         ?IMPLIES(M -- N == M andalso
                                  M -- lists:usort(M) == [],
                                  check_membership(M, N, D, B))))).

check_membership(M, N, D, B) ->
    {ok, Dlcbf} = new(D, B),
    F = lists:foldl(fun(X, Acc) ->
                            add(X, Acc),
                            Acc
                    end, Dlcbf, M),
    lists:all(fun(X) ->
                      in(X, F)
              end, M) and
        lists:all(fun(X) ->
                          not in(X, F)
                  end, N).

keys() ->
    eqc_gen:non_empty(list(eqc_gen:non_empty(binary()))).

ops(Keys) ->
    {oneof([add, delete]), oneof(Keys)}.

apply_ops([], _Dlcbf, Acc) ->
    Acc;
apply_ops([{add, Key}|T], Dlcbf, Acc) ->
    ok = dlcbf:add(Key, Dlcbf),
    apply_ops(T, Dlcbf, orddict:store(Key, ok, Acc));
apply_ops([{delete, Key}|T], Dlcbf, Acc) ->
    ok = dlcbf:delete(Key, Dlcbf),
    apply_ops(T, Dlcbf, orddict:store(Key, deleted, Acc)).

prop_add_delete() ->
    ?LET(Keys, keys(),
         ?FORALL(Ops0, eqc_gen:non_empty(list(ops(Keys))),
                 begin
                     {ok, Dlcbf} = new(4, 2048),
                     Ops = lists:usort(Ops0),
                     Model = apply_ops(Ops, Dlcbf, []),
                     F = fun({Key, deleted}) ->
                                 ?assertEqual(false, dlcbf:in(Key, Dlcbf));
                            ({Key, ok}) ->
                                 ?assertEqual(true, dlcbf:in(Key, Dlcbf))
                         end,
                     lists:map(F, Model),
                     true
                 end)).

prop_add_delete_test_() ->
    Prop = eqc:numtests(5000, dlcbf:prop_add_delete()),
    {timeout, 2*60, fun() -> qc(Prop) end}.

-endif. % EQC

-endif.
